snuba/core/initialize.py

```diff
@@ -0,0 +1,42 @@
+"""
+This file contains the initialization sequence for snuba. Anything that needs to be created before a snuba command
+is able to run is defined in this file under the `initialize` function
+
+
+All imports are hidden behind their respective functions to avoid running any code at initialization time unless explicitly
+directed
+"""
+import logging
+
+logger = logging.getLogger("snuba_core_init")
+
+
+def _load_datasets() -> None:
+    from snuba.datasets.factory import get_enabled_dataset_names
+
+    get_enabled_dataset_names()
+
+
+def _load_storages() -> None:
+    from snuba.datasets.storages.factory import get_all_storage_keys
+
+    get_all_storage_keys()
+
+
+def _load_entities() -> None:
+    from snuba.datasets.entities.factory import get_all_entity_names
+
+    get_all_entity_names()
+
+
+def initialize() -> None:
+    logger.info("Initializing snuba")
+
+    # The order of the functions matters. At time of writing (2022-10-27), snuba datasets are not guaranteed to load
+    # entities and storages. This is because datasets defined in code reference entities via EntityKey(s) and are
+    # thus loaded lazily
```

Comment:
I would remove this. Or consider rewording. This comment doesn't help me understand what's actually happening here.

Specifically from reading this function it looks like the entire storage, entity and dataset objects are being loaded. But then when looking at the implementations of each of the functions it looks like only the "names" or "keys' are being loaded, e.g. get_all_entity_names, get_all_storage_keys, etc, which implies that we might just be loading a small list of names or something. Can you clarify what's happening exactly?

Reply:
You're right. The functions I was using to initialize the factories are confusing. I used their proper init methods now. Let me know if it still doesn't make sense

Reply:
I'm not sure the reason provided here makes sense. Fundamentally the order matters because entities reference storages, so the storages need to be built first, and datasets reference entities etc.

In other words, if we changed the code so that datasets didn't load entities lazily, we would still probably want to maintain the bottom up ordering of initialization. We also want to be more transparent about what is being initialized when, so just loading the dataset and having everything else be initialized implicitly is counter to the benefits of this change.

Reply:
Updated the comment. Let me know if that makes sense